realvnc/configValuesFile.go

```diff
@@ -0,0 +1,122 @@
+//go:build !windows
+// +build !windows
+
+package realvnc
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+)
+
+var (
+	ErrMissingConfigValueName  = errors.New("missing config value name")
+	ErrConfigWriterCannotBeNil = errors.New("config writer cannot be nil")
+)
+
+type ConfigValue struct {
+	Name  string
+	Value string
+}
+
+type ConfigValues struct {
+	scanner      *bufio.Scanner
+	configWriter io.Writer
+}
+
+func NewConfigValuesFromFile(configFilename string) (configValues *ConfigValues, err error) {
+	configBytes, err := os.ReadFile(configFilename)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			// if we didn't find the file then still return valid config values with the err.
+			// we'll still be able to proceed with adding config values to a new file.
+			return &ConfigValues{}, err
+		}
+		return nil, err
+	}
+
+	configReader := bytes.NewReader(configBytes)
+	configValues, err = NewConfigValuesFromReader(configReader)
+	if err != nil {
+		return nil, err
+	}
+
+	return configValues, nil
+}
+
+func NewConfigValuesFromReader(reader io.Reader) (configValues *ConfigValues, err error) {
+	scanner := bufio.NewScanner(reader)
+
+	configValues = &ConfigValues{
+		scanner: scanner,
+	}
+
+	return configValues, nil
+}
+
+func (cv *ConfigValues) HasScanner() bool {
+	return cv.scanner != nil
+}
+
+func (cv *ConfigValues) Scan() bool {
+	return cv.scanner.Scan()
+}
+
+func (cv *ConfigValues) Text() string {
+	return cv.scanner.Text()
+}
+
+func (cv *ConfigValues) SetOutputWriter(writer io.Writer) {
+	cv.configWriter = writer
+}
+
+func (cv *ConfigValues) GetOutputWriter() (writer io.Writer) {
+	return cv.configWriter
+}
+
+func (cv *ConfigValues) WriteValue(configValue ConfigValue) (err error) {
+	outputLine := fmt.Sprintf("%s=%s", configValue.Name, configValue.Value)
+	return cv.WriteLine(outputLine)
+}
+
+func (cv *ConfigValues) WriteLine(outputLine string) (err error) {
+	if cv.configWriter == nil {
+		return ErrConfigWriterCannotBeNil
+	}
+	_, err = cv.configWriter.Write([]byte(outputLine + "\n"))
+	return err
+}
+
+func ApplyConfigChange(existingValue ConfigValue, inputValue ConfigValue) (
+	updated bool, updatedVal ConfigValue, err error) {
+	if inputValue.Value == existingValue.Value {
+		return false, existingValue, nil
+	}
+	return true, inputValue, nil
+}
+
+func ParseConfigKeyValueLine(inputLine string) (skip bool, configValue ConfigValue, err error) {
```

Comment:
I feel this is a bit difficult to use. I think instead of `ConfigValues` having `Text` method that then needs to be parsed with this function, it should have a `ConfigValue` method that returns this values and internally parses scanner text output.

Reply:
not sure i'm completely happy with the `ConfigValues` abstraction because it is a little weird. it needs to handle the plain lines of text as well as the key value pairs. the `ParseConfigKeyValueLine` is currently very easy to test and not coupled with anything else. i did look into making it a member fn but it just seemed to complicate the tests and i decided against.

so i prefer to leave as is this time? but also happy to ponder next time when revisiting the config file parsing.

Reply:
ok :+1: